src/components/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,140 @@
+  "aria-label": string
+  searchIcon?: ReactNode
+  clearIcon?: ReactNode
+  debounceDelay?: number
+  initialValue?: string
+  onSearch: (searchTerm: string) => void
+  showSearchButton?: boolean
+}
+
+export function SearchInput({
+  searchIcon = <span>🔍</span>,
+  clearIcon = <span>❎</span>,
+  onSearch,
+  debounceDelay = DEFAULT_DELAY,
+  showSearchButton,
+  initialValue = "",
+  ...inputProps
+}: SearchInputProps) {
+  const [searchTerm, setSearchTerm] = useState("")
+  const [clearButtonHidden, setClearButtonHidden] = useState(true)
+
+  const debouncedOnSearch = useMemo(() => {
+    return debounce(onSearch, debounceDelay)
+  }, [debounceDelay, onSearch])
+
+  useKeyPress(["Enter"], () => {
+    debouncedOnSearch(searchTerm)
+  })
+
+  useEffect(() => {
+    setSearchTerm(initialValue)
+  }, [initialValue])
+
+  const handleReset = (e: MouseEvent<HTMLButtonElement>) => {
+    e.preventDefault()
+    setSearchTerm("")
+    onSearch?.("")
+    setClearButtonHidden(true)
+  }
+
+  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const currentValue = e.currentTarget?.value
+    setSearchTerm(currentValue)
+    setClearButtonHidden(!currentValue)
+    shouldDebounceOnChange(showSearchButton) && debouncedOnSearch(currentValue)
+  }
+
+  const handleBlur = (e: FocusEvent<HTMLInputElement>) => {
+    inputProps.onBlur?.(e)
+    setClearButtonHidden(true)
+  }
+
+  const handleFocus = (e: FocusEvent<HTMLInputElement>) => {
+    inputProps.onFocus?.(e)
+    setClearButtonHidden(!searchTerm.length)
+  }
+
+  const clearButtonClasses = [
+    styles.clearButton,
+    clearButtonHidden && styles.hidden,
+  ].join(" ")
+
+  const inputClasses = [styles.input, inputProps.className].join(" ")
+
+  return (
+    <div className={styles.wrapper}>
+      <div className={styles.container} data-testid="container">
+        <span data-testid="search_icon" className={styles.searchIcon}>
+          {searchIcon}
+        </span>
+        <input
+          {...inputProps}
+          type="search"
+          className={inputClasses}
+          onBlur={handleBlur}
+          onFocus={handleFocus}
+          value={searchTerm}
+          onChange={handleChange}
+        />
+        <button
+          // we are ignoring tabbing since 'esc' is the keyboard accessible means of clearing an input
+          tabIndex={-1}
+          aria-label="click icon to clear search"
+          className={clearButtonClasses}
+          onMouseDown={handleReset}
+        >
+          {clearIcon}
+        </button>
+      </div>
+      {showSearchButton && (
+        <div className={styles.searchButton}>
+          <button
+            aria-label="click to search"
+            onClick={() => debouncedOnSearch(searchTerm)}
+          >
+            Search
+          </button>
+        </div>
+      )}
+    </div>
+  )
+}
+
+const shouldDebounceOnChange = (showSearchButton?: boolean) => {
+  return !showSearchButton
+}
```

Comment:
Looking at this function in isolation, it's not easy to see why it's there.

- First, an easy one, this function is a negation of its input, it could've been a statement with less overhead.

- Second, and most importantly, the name isn't clearly related to its input or logic. There's an implicit connection between the function name and debounce here, but what if we removed debounce from this component, would this function still exist? If so, then this function would need to be renamed, to what? And how does that related with showing the search button (aka, the input)?

This function (or variable if you make it an statement) is helping to fulfill a much higher-level concept:
> If there is a search button present on the screen, we don't want changes in the search value to automatically trigger an action `onSearch`. And then the opposite logic: if there isn't a search button, changes to the search value _should_ trigger the `onSearch` action.

And this logic is independent of debounce, so the challenge here is to pack all of this ☝🏼 into a single function/variable name.